app/queries/vacancy_filter_query.rb

```diff
@@ -51,6 +52,25 @@ def add_organisation_type_filters(filters, built_scope)
     built_scope.joins(:organisation_vacancies).where(organisation_vacancies: { vacancy_id: subquery })
   end
 
+  def add_school_type_filters(filters, built_scope)
+    return built_scope unless filters[:school_types].present?
+
+    subquery = school_type_filters_subquery(filters)
+
+    built_scope.joins(:organisation_vacancies).where(organisation_vacancies: { vacancy_id: subquery })
+  end
+
+  def school_type_filters_subquery(filters)
+    if filters[:school_types].include?("faith_school") && filters[:school_types].include?("special_school")
+      return OrganisationVacancy.joins(:organisation).where.not("organisations.gias_data ->> 'ReligiousCharacter (name)' IN (?)", Organisation::NON_FAITH_RELIGIOUS_CHARACTER_TYPES)
+                                                                             .or(OrganisationVacancy.where("organisations.school_type IN (?)", Organisation::SPECIAL_SCHOOL_TYPES)).select(:vacancy_id)
+    end
+
+    return OrganisationVacancy.joins(:organisation).where(organisations: { school_type: Organisation::SPECIAL_SCHOOL_TYPES }).select(:vacancy_id) if filters[:school_types].include?("special_school")
+
+    OrganisationVacancy.joins(:organisation).where.not("organisations.gias_data ->> 'ReligiousCharacter (name)' IN (?)", Organisation::NON_FAITH_RELIGIOUS_CHARACTER_TYPES).select(:vacancy_id)
+  end
```

Comment:
I have struggled to understand the logic for the conditions/queries here, so I checked this locally and rearranged it.
It composes the query bit by bit based on the conditions, leaving the common parts (and query triggering) for the end. 
This is clearer for me, but I leave it up to you to consider if you like it.

```suggestion
   def school_type_filters_subquery(school_types)
    special_schools_query = OrganisationVacancy.where(organisations: { school_type: Organisation::SPECIAL_SCHOOL_TYPES })
    query = OrganisationVacancy

    if school_types.include?("faith_school")
      query = query.where.not("organisations.gias_data ->> 'ReligiousCharacter (name)' IN (?)", Organisation::NON_FAITH_RELIGIOUS_CHARACTER_TYPES)
      query = query.or(special_schools_query) if school_types.include?("special_school")
    elsif school_types.include?("special_school")
      query = special_schools_query
    end
    query.joins(:organisation).select(:vacancy_id)
  end
```